File: problems/3381_maximum_subarray_sum_with_length_divisible_by_k/3381_maximum_subarray_sum_with_length_divisible_by_k.py
class Solution:
    def maxSubarraySum(self, nums: List[int], k: int) -> int:
        min_prefix = [float('inf')] * k
        min_prefix[k - 1] = 0
        s = 0
        ans = float('-inf')
        for i, x in enumerate(nums):
            s += x
            r = i % k
            ans = max(ans, s - min_prefix[r])
            min_prefix[r] = min(min_prefix[r], s)
        return ans

File: problems/3381_maximum_subarray_sum_with_length_divisible_by_k/3381_maximum_subarray_sum_with_length_divisible_by_k.java
class Solution {
    public long maxSubarraySum(int[] nums, int k) {
        long[] minPrefix = new long[k];
        final long INF = Long.MAX_VALUE / 2;
        Arrays.fill(minPrefix, INF);
        minPrefix[k - 1] = 0;
        long prefix = 0;
        long ans = Long.MIN_VALUE;
        for (int i = 0; i < nums.length; i++) {
            prefix += nums[i];
            int r = i % k;
            ans = Math.max(ans, prefix - minPrefix[r]);
            minPrefix[r] = Math.min(minPrefix[r], prefix);
        }
        return ans;
    }
}